# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [1]:
# Import the appropriate libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import style

# Create two new dataframes
north_df = pd.read_csv('NorthCallCenter.csv')
south_df = pd.read_csv('SouthCallCenter.csv')

## Exploratory Data Analysis

Time to do some EDA! In the process of learning more about the two datasets, make sure you answer the following questions:

1. How many reps are in each branch?
2. What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch?
3. What is the number of calls taken for each time block?

In [2]:
# Put your EDA code here
north_df

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0             0  North        Complaint    9:00 AM             Incoming     B   
1             1  North    Sales Support   11:00 AM             Incoming     A   
2             2  North    Sales Support   11:00 AM             Incoming     B   
3             3  North    Sales Support    4:00 PM             Incoming     B   
4             4  North  Product Support   12:00 PM             Incoming     B   
..          ...    ...              ...        ...                  ...   ...   
240         240  North  Product Support    5:00 PM             Outgoing     B   
241         241  North  Product Support    5:00 PM             Outgoing     A   
242         242  North  Product Support    5:00 PM             Outgoing     A   
243         243  North  Product Support    5:00 PM             Outgoing     A   
244         244  North  Product Support    5:00 PM             Outgoing     B   

     Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     Brent   NO          0          1             9                  2   
1       Cam   NO          0          0            11                  1   
2      Todd   NO          0          0             9                  3   
3     Brent   NO          0          0            11                  2   
4     Brent   NO          0          0             8                  2   
..      ...  ...        ...        ...           ...                ...   
240   Brent   NO          0          0            10                  2   
241     Joe   NO          0          0             8                  3   
242  Amanda   NO          0          0            12                  1   
243     Cam   NO          0          0            11                  2   
244  Xander   NO          0          0             8                  1   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      4      1  
1                   1.0                      3      1  
2                   1.0                      3      1  
3                   1.0                      4      1  
4                   1.0                      4      1  
..                  ...                    ...    ...  
240                 NaN                      3      1  
241                 NaN                      2      1  
242                 NaN                      2      1  
243                 NaN                      2      1  
244                 NaN                      4      1  

[245 rows x 15 columns]

In [3]:
# get columns with data type and non-null information
north_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245 entries, 0 to 244
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             245 non-null    int64  
 1   Branch                 245 non-null    object 
 2   Call Purpose           245 non-null    object 
 3   Time Block             245 non-null    object 
 4   Incoming or Outgoing   245 non-null    object 
 5   Queue                  245 non-null    object 
 6   Rep ID                 245 non-null    object 
 7   Sale                   245 non-null    object 
 8   Lost Call              245 non-null    int64  
 9   Abandoned              245 non-null    int64  
 10  Busy Minutes           245 non-null    int64  
 11  Not Ready Minutes      245 non-null    int64  
 12  Incoming Wait Time     82 non-null     float64
 13  During Call Wait Time  245 non-null    int64  
 14  Calls                  245 non-null    int64  
dtypes: flo

In [4]:
# how many reps are in each branch
north_reps = north_df['Rep ID'].value_counts()
len(north_reps)

9

In [5]:
# What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for north branch?
# be sure to exclude null values
north_ave = north_df.mean(axis = 0, skipna = True)
north_ave

C:\Users\tduen\AppData\Local\Temp/ipykernel_15344/4239815744.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  north_ave = north_df.mean(axis = 0, skipna = True)


Unnamed: 0               122.000000
Lost Call                  0.020408
Abandoned                  0.028571
Busy Minutes               9.987755
Not Ready Minutes          1.914286
Incoming Wait Time         3.048780
During Call Wait Time      2.967347
Calls                      1.000000
dtype: float64

In [6]:
# What is the number of calls taken for each time block for the north center?
north_time_blocks = north_df['Time Block'].value_counts()
north_time_blocks

12:00 PM    45
10:00 AM    43
2:00 PM     35
3:00 PM     30
4:00 PM     28
11:00 AM    25
1:00 PM     17
5:00 PM     17
9:00 AM      5
Name: Time Block, dtype: int64

In [7]:
south_df

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0             0  South    Sales Support   10:00 AM             Incoming     D   
1             1  South    Sales Support   10:00 AM             Incoming     C   
2             2  South    Sales Support   10:00 AM             Incoming     C   
3             3  South    Sales Support   10:00 AM             Incoming     C   
4             4  South    Sales Support   12:00 PM             Incoming     C   
..          ...    ...              ...        ...                  ...   ...   
309         309  South  Product Support    5:00 PM             Outgoing     D   
310         310  South  Product Support    5:00 PM             Outgoing     C   
311         311  South  Product Support    5:00 PM             Outgoing     C   
312         312  South  Product Support    5:00 PM             Outgoing     C   
313         313  South  Product Support    5:00 PM             Outgoing     D   

     Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0      Kate   NO          0          0             9                  1   
1      Eric   NO          0          0             8                  2   
2     Susan   NO          0          0            10                  2   
3     Alice   NO          0          0            12                  1   
4     Sandy   NO          0          0             8                  3   
..      ...  ...        ...        ...           ...                ...   
309   Helga   NO          0          0            10                  3   
310   Susan   NO          0          0            12                  3   
311   Sandy   NO          0          0             9                  3   
312    Eric   NO          0          0            10                  1   
313  George   NO          0          0            10                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      2      1  
1                   1.0                      4      1  
2                   1.0                      4      1  
3                   1.0                      3      1  
4                   1.0                      3      1  
..                  ...                    ...    ...  
309                 NaN                      4      1  
310                 NaN                      4      1  
311                 NaN                      4      1  
312                 NaN                      2      1  
313                 NaN                      3      1  

[314 rows x 15 columns]

In [8]:
# get columns with data type and non-null information
south_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 314 entries, 0 to 313
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             314 non-null    int64  
 1   Branch                 314 non-null    object 
 2   Call Purpose           314 non-null    object 
 3   Time Block             314 non-null    object 
 4   Incoming or Outgoing   314 non-null    object 
 5   Queue                  314 non-null    object 
 6   Rep ID                 314 non-null    object 
 7   Sale                   314 non-null    object 
 8   Lost Call              314 non-null    int64  
 9   Abandoned              314 non-null    int64  
 10  Busy Minutes           314 non-null    int64  
 11  Not Ready Minutes      314 non-null    int64  
 12  Incoming Wait Time     126 non-null    float64
 13  During Call Wait Time  314 non-null    int64  
 14  Calls                  314 non-null    int64  
dtypes: flo

In [9]:
# how many reps are in each branch
south_reps = south_df['Rep ID'].value_counts()
len(south_reps)

11

In [10]:
# What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for south branch?
# be sure to exclude null values
south_ave = south_df.mean(axis = 0, skipna = True)
south_ave

C:\Users\tduen\AppData\Local\Temp/ipykernel_15344/2945674402.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  south_ave = south_df.mean(axis = 0, skipna = True)


Unnamed: 0               156.500000
Lost Call                  0.035032
Abandoned                  0.012739
Busy Minutes              10.054140
Not Ready Minutes          1.910828
Incoming Wait Time         3.000000
During Call Wait Time      3.076433
Calls                      1.000000
dtype: float64

In [11]:
# What is the number of calls taken for each time block for the south center?
south_time_blocks = south_df['Time Block'].value_counts()
south_time_blocks

12:00 PM    75
10:00 AM    56
3:00 PM     43
11:00 AM    31
2:00 PM     30
5:00 PM     26
4:00 PM     25
1:00 PM     23
9:00 AM      5
Name: Time Block, dtype: int64

### EDA Results

Make note of your results here!

1. How many reps are in each branch? 

North = 9 

South = 11

2. What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch?

**North**

Busy Minutes               9.987755

Not Ready Minutes          1.914286

Incoming Wait Time         3.048780

During Call Wait Time      2.967347

**South**

Busy Minutes              10.054140

Not Ready Minutes          1.910828

Incoming Wait Time         3.000000

During Call Wait Time      3.076433

3. What is the number of calls taken for each time block?

**North**

12:00 PM    45

10:00 AM    43

2:00 PM     35

3:00 PM     30

4:00 PM     28

11:00 AM    25

1:00 PM     17

5:00 PM     17

9:00 AM      5

**South**

12:00 PM    75

10:00 AM    56

3:00 PM     43

11:00 AM    31

2:00 PM     30

5:00 PM     26

4:00 PM     25

1:00 PM     23

9:00 AM      5

## Cleaning Data

Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets:

1. Incoming wait time is null in many places. Do all of those places correspond to outgoing calls?
2. In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
3. Are there duplicates in the two data sets? If there are, how did you handle them?
4. Is any of the data in the two data sets unnecessary? If yes, how did you handle it?

In [12]:
# Clean your data here
# north_null = north_df[north_df['Incoming Wait Time'] == 'NaN']['Incoming or Outgoing']
# north_null

#north_df.loc[north_df['Incoming Wait Time'].isin(["NaN"])]

north_df.loc[(north_df['Incoming or Outgoing'] == 'Outgoing') & (north_df['Incoming Wait Time'] == 3.0)]

Empty DataFrame
Columns: [Unnamed: 0, Branch, Call Purpose, Time Block, Incoming or Outgoing, Queue, Rep ID, Sale, Lost Call, Abandoned, Busy Minutes, Not Ready Minutes, Incoming Wait Time, During Call Wait Time, Calls]
Index: []

In [13]:
# Clean your data here
# north_null = north_df[north_df['Incoming Wait Time'] == 'NaN']['Incoming or Outgoing']
# north_null

#north_df.loc[north_df['Incoming Wait Time'].isin(["NaN"])]

#incoming_north_null = north_df.filter(items = ['Incoming or Outgoing','Incoming Wait Time'])
incoming_north_null = north_df[north_df['Incoming or Outgoing'] == 'Incoming']
incoming_north_null.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 82 entries, 0 to 81
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             82 non-null     int64  
 1   Branch                 82 non-null     object 
 2   Call Purpose           82 non-null     object 
 3   Time Block             82 non-null     object 
 4   Incoming or Outgoing   82 non-null     object 
 5   Queue                  82 non-null     object 
 6   Rep ID                 82 non-null     object 
 7   Sale                   82 non-null     object 
 8   Lost Call              82 non-null     int64  
 9   Abandoned              82 non-null     int64  
 10  Busy Minutes           82 non-null     int64  
 11  Not Ready Minutes      82 non-null     int64  
 12  Incoming Wait Time     82 non-null     float64
 13  During Call Wait Time  82 non-null     int64  
 14  Calls                  82 non-null     int64  
dtypes: float

In [14]:
north_df['Sale'].unique().tolist()

['NO', 'YES ', 'YES']

In [20]:
# 2. two separate 'Yes' values.

north_df.loc[north_df['Sale'].isin(['YES '])]
north_df['Sale'] = north_df['Sale'].replace(['YES '],'YES')
north_df.head(20)

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0            0  North        Complaint    9:00 AM             Incoming     B   
1            1  North    Sales Support   11:00 AM             Incoming     A   
2            2  North    Sales Support   11:00 AM             Incoming     B   
3            3  North    Sales Support    4:00 PM             Incoming     B   
4            4  North  Product Support   12:00 PM             Incoming     B   
5            5  North  Product Support    2:00 PM             Incoming     B   
6            6  North  Product Support    2:00 PM             Incoming     A   
7            7  North  Product Support    2:00 PM             Incoming     B   
8            8  North  Product Support    3:00 PM             Incoming     B   
9            9  North  Product Support    3:00 PM             Incoming     A   
10          10  North  Product Support    4:00 PM             Incoming     A   
11          11  North  Product Support    4:00 PM             Incoming     A   
12          12  North  Product Support    4:00 PM             Incoming     B   
13          13  North        Complaint    9:00 AM             Incoming     B   
14          14  North        Complaint   12:00 PM             Incoming     B   
15          15  North        Complaint    1:00 PM             Incoming     B   
16          16  North        Complaint    3:00 PM             Incoming     B   
17          17  North        Complaint    2:00 PM             Incoming     A   
18          18  North    Sales Support   10:00 AM             Incoming     B   
19          19  North    Sales Support   10:00 AM             Incoming     B   

    Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0    Brent   NO          0          1             9                  2   
1      Cam   NO          0          0            11                  1   
2     Todd   NO          0          0             9                  3   
3    Brent   NO          0          0            11                  2   
4    Brent   NO          0          0             8                  2   
5     Duke   NO          0          0            12                  2   
6      Joe   NO          0          0             8                  1   
7    Lilly   NO          0          0            11                  2   
8     Duke   NO          0          0            12                  2   
9      Cam   NO          0          0            11                  3   
10     Joe   NO          0          0            11                  3   
11  Amanda   NO          0          0            10                  3   
12    Todd   NO          0          0            12                  2   
13    Duke   NO          0          0             9                  2   
14   Brent   NO          0          0             8                  2   
15    Duke   NO          0          0            12                  1   
16   Brent   NO          0          0             9                  3   
17    Andy   NO          0          1             9                  3   
18    Duke   NO          0          0            12                  2   
19   Lilly  YES          0          0            11                  2   

    Incoming Wait Time  During Call Wait Time  Calls  
0                  1.0                      4      1  
1                  1.0                      3      1  
2                  1.0                      3      1  
3                  1.0                      4      1  
4                  1.0                      4      1  
5                  1.0                      2      1  
6                  1.0                      3      1  
7                  1.0                      3      1  
8                  1.0                      4      1  
9                  1.0                      4      1  
10                 1.0                      4      1  
11                 1.0                      2      1  
12                 1.0                      3      1 

In [40]:
north_dup_check = north_df.drop(['Unnamed: 0'], axis = 1)
#north_dup_check.duplicated(keep = False)
north_dup_check

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    North        Complaint    9:00 AM             Incoming     B   Brent   
1    North    Sales Support   11:00 AM             Incoming     A     Cam   
2    North    Sales Support   11:00 AM             Incoming     B    Todd   
3    North    Sales Support    4:00 PM             Incoming     B   Brent   
4    North  Product Support   12:00 PM             Incoming     B   Brent   
..     ...              ...        ...                  ...   ...     ...   
240  North  Product Support    5:00 PM             Outgoing     B   Brent   
241  North  Product Support    5:00 PM             Outgoing     A     Joe   
242  North  Product Support    5:00 PM             Outgoing     A  Amanda   
243  North  Product Support    5:00 PM             Outgoing     A     Cam   
244  North  Product Support    5:00 PM             Outgoing     B  Xander   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO          0          1             9                  2   
1     NO          0          0            11                  1   
2     NO          0          0             9                  3   
3     NO          0          0            11                  2   
4     NO          0          0             8                  2   
..   ...        ...        ...           ...                ...   
240   NO          0          0            10                  2   
241   NO          0          0             8                  3   
242   NO          0          0            12                  1   
243   NO          0          0            11                  2   
244   NO          0          0             8                  1   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      4      1  
1                   1.0                      3      1  
2                   1.0                      3      1  
3                   1.0                      4      1  
4                   1.0                      4      1  
..                  ...                    ...    ...  
240                 NaN                      3      1  
241                 NaN                      2      1  
242                 NaN                      2      1  
243                 NaN                      2      1  
244                 NaN                      4      1  

[245 rows x 14 columns]

In [42]:
ndc = north_dup_check.drop_duplicates()
ndc

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    North        Complaint    9:00 AM             Incoming     B   Brent   
1    North    Sales Support   11:00 AM             Incoming     A     Cam   
2    North    Sales Support   11:00 AM             Incoming     B    Todd   
3    North    Sales Support    4:00 PM             Incoming     B   Brent   
4    North  Product Support   12:00 PM             Incoming     B   Brent   
..     ...              ...        ...                  ...   ...     ...   
240  North  Product Support    5:00 PM             Outgoing     B   Brent   
241  North  Product Support    5:00 PM             Outgoing     A     Joe   
242  North  Product Support    5:00 PM             Outgoing     A  Amanda   
243  North  Product Support    5:00 PM             Outgoing     A     Cam   
244  North  Product Support    5:00 PM             Outgoing     B  Xander   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO          0          1             9                  2   
1     NO          0          0            11                  1   
2     NO          0          0             9                  3   
3     NO          0          0            11                  2   
4     NO          0          0             8                  2   
..   ...        ...        ...           ...                ...   
240   NO          0          0            10                  2   
241   NO          0          0             8                  3   
242   NO          0          0            12                  1   
243   NO          0          0            11                  2   
244   NO          0          0             8                  1   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      4      1  
1                   1.0                      3      1  
2                   1.0                      3      1  
3                   1.0                      4      1  
4                   1.0                      4      1  
..                  ...                    ...    ...  
240                 NaN                      3      1  
241                 NaN                      2      1  
242                 NaN                      2      1  
243                 NaN                      2      1  
244                 NaN                      4      1  

[243 rows x 14 columns]

In [36]:
south_dup_check = south_df.drop(['Unnamed: 0'], axis = 1)
south_dup_check.duplicated(keep = 'last')


0      False
1      False
2      False
3      False
4      False
       ...  
309    False
310    False
311    False
312    False
313    False
Length: 314, dtype: bool

In [52]:
sdc = south_dup_check.drop_duplicates()
sdc

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    South    Sales Support   10:00 AM             Incoming     D    Kate   
1    South    Sales Support   10:00 AM             Incoming     C    Eric   
2    South    Sales Support   10:00 AM             Incoming     C   Susan   
3    South    Sales Support   10:00 AM             Incoming     C   Alice   
4    South    Sales Support   12:00 PM             Incoming     C   Sandy   
..     ...              ...        ...                  ...   ...     ...   
309  South  Product Support    5:00 PM             Outgoing     D   Helga   
310  South  Product Support    5:00 PM             Outgoing     C   Susan   
311  South  Product Support    5:00 PM             Outgoing     C   Sandy   
312  South  Product Support    5:00 PM             Outgoing     C    Eric   
313  South  Product Support    5:00 PM             Outgoing     D  George   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO          0          0             9                  1   
1     NO          0          0             8                  2   
2     NO          0          0            10                  2   
3     NO          0          0            12                  1   
4     NO          0          0             8                  3   
..   ...        ...        ...           ...                ...   
309   NO          0          0            10                  3   
310   NO          0          0            12                  3   
311   NO          0          0             9                  3   
312   NO          0          0            10                  1   
313   NO          0          0            10                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      2      1  
1                   1.0                      4      1  
2                   1.0                      4      1  
3                   1.0                      3      1  
4                   1.0                      3      1  
..                  ...                    ...    ...  
309                 NaN                      4      1  
310                 NaN                      4      1  
311                 NaN                      4      1  
312                 NaN                      2      1  
313                 NaN                      3      1  

[311 rows x 14 columns]

In [56]:
sdc.duplicated(['Calls'], keep = False)

0      True
1      True
2      True
3      True
4      True
       ... 
309    True
310    True
311    True
312    True
313    True
Length: 311, dtype: bool

In [58]:
north_clean = ndc.drop('Calls', axis = 1)
north_clean

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    North        Complaint    9:00 AM             Incoming     B   Brent   
1    North    Sales Support   11:00 AM             Incoming     A     Cam   
2    North    Sales Support   11:00 AM             Incoming     B    Todd   
3    North    Sales Support    4:00 PM             Incoming     B   Brent   
4    North  Product Support   12:00 PM             Incoming     B   Brent   
..     ...              ...        ...                  ...   ...     ...   
240  North  Product Support    5:00 PM             Outgoing     B   Brent   
241  North  Product Support    5:00 PM             Outgoing     A     Joe   
242  North  Product Support    5:00 PM             Outgoing     A  Amanda   
243  North  Product Support    5:00 PM             Outgoing     A     Cam   
244  North  Product Support    5:00 PM             Outgoing     B  Xander   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO          0          1             9                  2   
1     NO          0          0            11                  1   
2     NO          0          0             9                  3   
3     NO          0          0            11                  2   
4     NO          0          0             8                  2   
..   ...        ...        ...           ...                ...   
240   NO          0          0            10                  2   
241   NO          0          0             8                  3   
242   NO          0          0            12                  1   
243   NO          0          0            11                  2   
244   NO          0          0             8                  1   

     Incoming Wait Time  During Call Wait Time  
0                   1.0                      4  
1                   1.0                      3  
2                   1.0                      3  
3                   1.0                      4  
4                   1.0                      4  
..                  ...                    ...  
240                 NaN                      3  
241                 NaN                      2  
242                 NaN                      2  
243                 NaN                      2  
244                 NaN                      4  

[243 rows x 13 columns]

In [59]:
south_clean = sdc.drop('Calls', axis = 1)
south_clean 

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    South    Sales Support   10:00 AM             Incoming     D    Kate   
1    South    Sales Support   10:00 AM             Incoming     C    Eric   
2    South    Sales Support   10:00 AM             Incoming     C   Susan   
3    South    Sales Support   10:00 AM             Incoming     C   Alice   
4    South    Sales Support   12:00 PM             Incoming     C   Sandy   
..     ...              ...        ...                  ...   ...     ...   
309  South  Product Support    5:00 PM             Outgoing     D   Helga   
310  South  Product Support    5:00 PM             Outgoing     C   Susan   
311  South  Product Support    5:00 PM             Outgoing     C   Sandy   
312  South  Product Support    5:00 PM             Outgoing     C    Eric   
313  South  Product Support    5:00 PM             Outgoing     D  George   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO          0          0             9                  1   
1     NO          0          0             8                  2   
2     NO          0          0            10                  2   
3     NO          0          0            12                  1   
4     NO          0          0             8                  3   
..   ...        ...        ...           ...                ...   
309   NO          0          0            10                  3   
310   NO          0          0            12                  3   
311   NO          0          0             9                  3   
312   NO          0          0            10                  1   
313   NO          0          0            10                  2   

     Incoming Wait Time  During Call Wait Time  
0                   1.0                      2  
1                   1.0                      4  
2                   1.0                      4  
3                   1.0                      3  
4                   1.0                      3  
..                  ...                    ...  
309                 NaN                      4  
310                 NaN                      4  
311                 NaN                      4  
312                 NaN                      2  
313                 NaN                      3  

[311 rows x 13 columns]

### Cleaning Data Results

Make note of your results!

**1. Incoming wait time is null in many places. Do all of those places correspond to outgoing calls?**

Yes, I was able to determine that all of the null values correspond to outgoing calls. I created a separate data frame that looked specifically at the incoming data in 'Incoming or Outgoing" and what corresponded to those rows.

**2. In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?**

One would try to correct the values so as to not skew data.  I was able to find how many instances and used the replace function in pandas. 

**3. Are there duplicates in the two data sets? If there are, how did you handle them?**

Found duplicates in both by removing "Unnamed: 0" column. I used .drop_duplicates() to remove the duplicates. 

**4. Is any of the data in the two data sets unnecessary? If yes, how did you handle it?**

I removed the "Unnamed: 0" and "Calls" columns as they didn't add any insight. 

## Data Manipulation

Before you begin answering any questions, combine the two datasets together to create a third dataframe called `df`. You can use this third dataframe to compare the two call centers to company-wide trends.

1. Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.
2. The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?
3. Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?
4. For each branch and the entire company, what is the average speed of answer?

In [16]:
# Manipulate data here

### Data Manipulation Results

1. Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.
2. The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?
3. Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each company and the entire company?
4. For each branch and the entire company, what is the average speed of answer?

## Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why.

1. What is the average abandonment rate per queue?
2. What is the service level and average speed of answer per each rep?
3. For each type of call purpose, how many calls are outgoing vs. incoming?

In [17]:
# Create visualization 1 here
# The dictionary abandonment_rates has the data you need.

abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = df.groupby("Queue").agg("sum")
for i in range(4):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

NameError: name 'df' is not defined

In [ ]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = north.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = north[north["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = north[north["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

In [ ]:
# Create visualization 3 here
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

### Visualization Results

For each chart you created, explain why you chose the chart style you chose.

## Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.

1. Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why?
2. Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help?
3. Now that you have explored the datasets, is there any data or information that you wish you had in this analysis?

## Submit Your Work

After you have completed your work on the assignment, push your work to your Github repo. Navigate to the repo, copy the link to your assignment and paste the URL in the Canvas assignment's submission box. 

## Bonus Mission

Create a visualization that answers this question:

1. For each call purpose, how many calls (incoming and outgoing) take place in each time block?

In [ ]:
# Create your visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot:
